documentation/projects/proposals/search_relevancy_sandbox/20230406-implementation_plan_update_staging_database.md

```diff
@@ -0,0 +1,305 @@
+   be named in a way that does not conflict with the existing staging database
+   name, e.g. `dev-next-openverse-db`. The database configuration information
+   from the previous step will be used to ensure the new database matches the
+   old database's configuration exactly.
+6. Wait for the new database to be ready using the
+   [`RdsDbSensor`](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/sensors/rds/index.html#airflow.providers.amazon.aws.sensors.rds.RdsDbSensor)
+   ([example](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/operators/rds.html#howto-sensor-rdsdbsensor)).
+7. Rename the old staging database to `dev-old-openverse-db` using boto3's
+   [`modify_db_instance`](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/rds/client/modify_db_instance.html)
+   function (an operator does not exist for this operation). We must set the
+   `ApplyImmediately` option here to `True` to ensure this operation happens
+   immediately rather than waiting for the next maintenance window.
+8. Wait for the old database rename to be complete with the `RdsDbSensor` (we
+   may need retries on this step, since the database may not be initially
+   available/named when the sensor first starts). _**Note**: this will cause a
+   temporary outage of the staging API, see
+   [the alternatives section](#alternatives) for why this is the case._ A Slack
+   notification should be sent out at the start of this step to alert the team
+   of the outage, and again once the outage is resolved.
+9. Rename the new database to `dev-openverse-db` using `modify_db_instance`.
+   (Noting that `ApplyImmediately` should be set to `True` here as well.)
+10. Wait for the new database rename to be complete with the `RdsDbSensor` (we
+    may need retries on this step, since the database may not be initially
+    available/named when the sensor first starts).
+11. If the previous steps fail, rename `dev-old-openverse-db` back to
+    `dev-openverse-db`. Otherwise, `dev-old-openverse-db` can be deleted using
+    the
+    [`RdsDeleteDbInstanceOperator`](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/operators/rds/index.html#airflow.providers.amazon.aws.operators.rds.RdsDeleteDbInstanceOperator).
+    We should use the following configuration options:
+    - `wait_for_completion` should be left as `True` here so this stop will hang
+      until the database removal is complete.
+    - `SkipFinalSnapshot` should be set to `True` to avoid creating a final
+      snapshot of the database before deletion.
+12. Once the database rename is complete, the following actions will also need
+    to occur (they can occur simultaneously in the DAG):
+    1. The staging API must be deployed to apply any necessary migrations to the
+       new copy the database. In order to do this we must augment the existing
```

Comment:
If the requirement is to only run migrations it would be easier and faster just to run the django command `python manage.py migrate` in sstaging instance. Are there additional reasons that warrant a deployment?

Reply:
To run a command in staging we'd need to enable [AWS SSM](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-exec.html) on the AWS client installed in Airflow. ECS Exec only works if SSM is enabled on the client. I don't know how much work it would be to set that up for the catalog. Staging API deployments take  <=3 minutes. My 2 cents is that it probably isn't worth the hassle for this particular use-case.

Reply:
Yea agreed - I think a deployment would be the easiest thing to kick off given what we have.